File: 5.最长回文子串.py
#
# @lc app=leetcode.cn id=5 lang=python
#
# [5] 最长回文子串
#

# @lc code=start
class Solution(object):
    def expandAroundCenter(self, s, left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return left+1, right-1
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        start, end = 0, 0
        for i in range(len(s)):
            left1, right1 = self.expandAroundCenter(s, i, i )
            left2, right2 = self.expandAroundCenter(s, i, i+1)
            if right1 - left1 > end - start:
                end = right1
                start = left1
            if right2 - left2 > end - start:
                end = right2
                start = left2
        return s[start:end+1]

# @lc code=end

File: 5.最长回文子串.cpp
/*
 * @lc app=leetcode.cn id=5 lang=cpp
 *
 * [5] 最长回文子串
 */

// @lc code=start
class Solution {
public:
    string longestPalindrome(string s) {
        if(s.size() < 2)
            return s;
        string result = "";
        for(int i = 0; i < s.size(); i++)
        {
            int left = i, right = i;
            while(left > 0 and s[left-1] == s[i])
                left--;
            while(right < s.size()-1 and s[right+1] == s[i])
                right++;
            cout << left << " " << right << "\n";
            bool flag =false;
            while(left >=0 and right < s.size())
            {
                if(s[left] != s[right])
                {
                    flag = true;
                    break;
                }
                if(left == 0 or right == s.size()-1)
                    break;
                left--;
                right++;
            }
            if(flag)
            {
                left += 1;
                right -= 1;
            }
            string tmp = s.substr(left, right-left+1);
            if(tmp.size() > result.size())
                result = tmp;
        }
        return result;
    }
};
// @lc code=end

